File: Code/Chapter 6/C#/Example 6.3/Form1.cs
#region Using directives

using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Windows.Forms;

#endregion

namespace Example_6_3
{
   partial class Form1 : Form
   {
      private DataSet petsData;
      public Form1()
      {
         InitializeComponent();
      }

      private void btnPopulate_Click(object sender, EventArgs e)
      {
         petsData = CreateSchema();
         petsGrid.DataSource = petsData.Tables["Pet"];
         AnimalsGrid.DataSource = petsData.Tables["Animal"];
      }

      private DataSet CreateSchema()
      {
         DataTable AnimalTable = new DataTable("Animal");
         DataColumn myDataColumn;

         myDataColumn = new DataColumn("AnimalID", typeof(System.Int32));
         myDataColumn.AutoIncrement = true;
         AnimalTable.Columns.Add(myDataColumn);

         myDataColumn = new DataColumn("AnimalName", typeof(System.String));
         AnimalTable.Columns.Add(myDataColumn);

         DataTable petsTable = new DataTable("Pet");

         myDataColumn = new DataColumn("PetID", typeof(System.Int32));
         myDataColumn.AutoIncrement = true;
         petsTable.Columns.Add(myDataColumn);

         myDataColumn = new DataColumn("AnimalID", typeof(System.Int32));
         petsTable.Columns.Add(myDataColumn);

         myDataColumn = new DataColumn("PetName", typeof(System.String));
         petsTable.Columns.Add(myDataColumn);

         DataSet toReturn = new DataSet("petsData");
         toReturn.Tables.Add(AnimalTable);
         toReturn.Tables.Add(petsTable);
         toReturn.Relations.Add(
            new DataRelation("AnimalsPets",
            AnimalTable.Columns["AnimalID"], petsTable.Columns["AnimalID"]));

         return toReturn;
      }

      private void btnShowXml_Click(object sender, EventArgs e)
      {
         MessageBox.Show(petsData.GetXml(), "DataSet Contents");
      }

      private void nestedRelation_CheckedChanged(object sender, EventArgs e)
      {
         if (petsData != null)
         {
            petsData.Relations[0].Nested = nestedRelation.Checked;
         }
      }
   }
}

File: Code/Chapter 6/VB.NET/Exercise 6.3/Form1.vb
Imports System.Data

Public Class Form1

   Dim petsData As DataSet
   Private Sub btnPopulate_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles btnSetupGrids.Click
      petsData = CreateSchema()
      petsGrid.DataSource = petsData.Tables("Pet")
      AnimalsGrid.DataSource = petsData.Tables("Animal")
   End Sub
   Private Sub btnShowXml_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles btnShowXml.Click
      MessageBox.Show(petsData.GetXml(), "DataSet Contents")
   End Sub
   Private Sub nestedRelation_CheckedChanged(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles nestedRelation.CheckedChanged
      If Not petsData Is Nothing Then
         petsData.Relations(0).Nested = nestedRelation.Checked
      End If
   End Sub

   Private Function CreateSchema() As DataSet
      Dim AnimalTable As DataTable = New DataTable("Animal")
      Dim myDataColumn As DataColumn

      myDataColumn = New DataColumn("AnimalID", GetType(System.Int32))
      myDataColumn.AutoIncrement = True
      AnimalTable.Columns.Add(myDataColumn)

      myDataColumn = New DataColumn("AnimalName", GetType(System.String))
      AnimalTable.Columns.Add(myDataColumn)

      Dim petsTable As DataTable = New DataTable("Pet")

      myDataColumn = New DataColumn("PetID", GetType(System.Int32))
      myDataColumn.AutoIncrement = True
      petsTable.Columns.Add(myDataColumn)

      myDataColumn = New DataColumn("AnimalID", GetType(System.Int32))
      petsTable.Columns.Add(myDataColumn)

      myDataColumn = New DataColumn("PetName", GetType(System.String))
      petsTable.Columns.Add(myDataColumn)

      Dim toReturn As DataSet = New DataSet("petsData")
      toReturn.Tables.Add(AnimalTable)
      toReturn.Tables.Add(petsTable)
      toReturn.Relations.Add( _
         New DataRelation("AnimalsPets", _
         AnimalTable.Columns("AnimalID"), petsTable.Columns("AnimalID")))

      Return toReturn
   End Function
End Class
